Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+        } catch (e: Exception) {
+            Timber.tag(TAG).e("Failed to register public key for %s", cert)
+            throw e
+        }
+    }
+
+    private suspend fun obtainCertificate(
+        cert: TestCertificateContainer
+    ): TestCertificateContainer {
+        return try {
+            Timber.tag(TAG).d("requestCertificate(cert=%s)", cert)
+
+            if (!cert.isPublicKeyRegistered) throw IllegalStateException("Public key is not registered yet.")
+
+            if (!cert.isPending) {
+                Timber.tag(TAG).d("Dcc has already been retrieved for %s", cert)
+                return cert
+            }
+
+            val certConfig = appConfigProvider.currentConfig.first().covidCertificateParameters.testCertificate
+
+            val nowUTC = timeStamper.nowUTC
+            val certAvailableAt = cert.publicKeyRegisteredAt!!.plus(certConfig.waitAfterPublicKeyRegistration)
+            val certAvailableIn = Duration(nowUTC, certAvailableAt)
+
+            val components = withContext(dispatcherProvider.IO) {
+                if (certAvailableIn > Duration.ZERO && certAvailableIn < certConfig.waitAfterPublicKeyRegistration) {
+                    Timber.tag(TAG).d("Delaying certificate retrieval by %d ms", certAvailableIn.millis)
+                    delay(certAvailableIn.millis)
+                }
+
+                val executeRequest: suspend CoroutineScope.() -> TestCertificateComponents = {
+                    certificateServer.requestCertificateForTest(testRegistrationToken = cert.registrationToken)
+                }
+
+                try {
+                    executeRequest()
+                } catch (e: Exception) {
+                    // TODO catch a specific error that reflects error code DGC_COMP_202
+                    delay(certConfig.waitForRetry.millis)
+                    executeRequest()
+                }
+            }
+            Timber.tag(TAG).i("Test certificate components successfully request for %s: %s", cert, components)
+
+            val encryptionkey = rsaCryptography.decrypt(
+                toDecrypt = components.dataEncryptionKeyBase64.decodeBase64()!!,
+                privateKey = cert.rsaPrivateKey!!
+            )
+
+            val extractedData = qrCodeExtractor.extract(
+                decryptionKey = encryptionkey.toByteArray(),
+                encryptedCoseComponents = components.encryptedCoseTestCertificateBase64.decodeBase64()!!
+            )
+
+            val nowUtc = timeStamper.nowUTC
+
+            when (cert.type) {
+                CoronaTest.Type.PCR -> (cert as PCRCertificateContainer).copy(
+                    testCertificateQrCode = extractedData.qrCode,
+                    certificateReceivedAt = nowUtc,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> (cert as RACertificateContainer).copy(
+                    testCertificateQrCode = extractedData.qrCode,
+                    certificateReceivedAt = nowUtc,
+                )
```

Comment:
would that work?
```kotlin                  
                   when (cert) {
                        is PCRCertificateContainer -> it.copy()
                        is RACertificateContainer -> it.copy()
                    }
```

Reply:
Yes, but then we would also need an `else ->`, while less, if someone adds a 3rd test type, we would not automatically get a compiler error here to notify us that we need to add extra handling here.

While less code, I think it's easier for future developers to add new test types if the compiler helps to fix things like this.